src/streams.lisp

```diff
@@ -183,6 +215,7 @@
     (when (plusp fill-ptr)
       (unless handle
         (error "output operation on closed SSL stream"))
+      (set-deadline-from-timeout stream :output)
```

Comment:
This approach, where every Gray Streams generic functions computes a deadline and stores it as the stream slot, which is later used to decide the timeout expiration, worries me somewhat. It introduces a requirement, that every Gray Stream generic function that my call `ensure-ssl-funcal` has to remember computing and setting the fresh deadline. I worry that in the future code modifications some code path may miss setting new deadline, and the old deadline will be in effect, signalling unnecessary error.

Your approach is probably motivated by the desire to apply the timeout to the full gray stream function, as opposed to smaller scope "any piece of data received". However, even in the current version of the code, the `stream-write-sequence` implementation can apply timeout to chunks of data, instead of the whole data.

I see the following ways to make the implementation more stable in the face of the future changes:
- instead of a stream slot, make the deadline a local variable in the `io-wait` or `ensure-ssl-funcall`, thus shifting the timeout semantics from "the whole stream function completion" to "any data received"
- introduce a macro `(with-deadline-from-timeout (stream) ...)` that will restore the `deadline` field to special value `none` upon exit, and upon enter will only modify the `deadline` slot if it's current value is `none`, which will make nested calls to `(with-deadline-from-timeout (stream)...) a no-operation. It will allow to apply the timeout to the whole `write-sequence`, while still having a timeout for `force-output` in case it's called by the application directly.

The choice depends on which timeout semantics is preferred. 

What do you think, what is better: timeout means the time of full completion of a function, like `read-sequence`, or to mean the time to wait for any data?

FWIW, the SBCL's function `socket-make-stream ` uses the latter behavior.

Reply:
I agree things are nicer with "timeout is the time to wait for any data".  I think your worry and suggestions are both good and will do them.